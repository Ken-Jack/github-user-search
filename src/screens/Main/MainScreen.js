// @flow

import React from 'react';
import qs from 'query-string';
import {Route} from 'react-router-dom';
import curry from 'lodash/fp/curry';
import DefaultLayout from 'components/DefaultLayout/DefaultLayout';
import {PER_PAGE} from 'store/api';

type Props = {
  component: Function,
};

export const pushUrlQuery = curry((push: Function, query: string) => {
  push({
    pathname: '/search',
    search: `per_page=${PER_PAGE}&page=1&q=${query}`,
  });
});

export default function MainScreen(props: Props) {
  const {
    component,
    ...restProps
  } = props;
  return (
    <Route
      {...restProps}
      render={matchProps => {
        const {
          location,
          history: {push},
        } = matchProps;
        const parsedSearch = qs.parse(location.search);
        const searchTerm = parsedSearch.q;

        return (
          <DefaultLayout
            onSearchSubmit={pushUrlQuery(push)}
            searchTerm={searchTerm}
            searchQuery={qs.stringify(parsedSearch)}
            component={component}
            matchProps={matchProps}
          />
        );
      }}
    />

  );
}
