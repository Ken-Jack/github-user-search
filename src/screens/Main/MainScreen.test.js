import React from 'react';
import renderer from 'react-test-renderer';
import {Provider} from 'react-redux';
import {
  StaticRouter as Router,
} from 'react-router-dom';
import configureMockStore from 'redux-mock-store';
import MainScreen, {
  pushUrlQuery,
} from './MainScreen';

function createComponent(component) {
  const mockStore = configureMockStore();
  const store = mockStore();
  return renderer.create(
    <Provider store={store}>
      <Router context={{}}>
        {component}
      </Router>
    </Provider>
  ).toJSON();
}

describe('Screen: MainScreen', () => {

  describe('when passed a component', () => {
    it('should render it in the default layout', () => {
      const component = () => <p>testing</p>;
      const screen = <MainScreen
        exact={true}
        path="/search"
        component={component}
        />
      expect(createComponent(screen)).toMatchSnapshot();
    });
  });

  describe('pushUrlQuery function', () => {
    it('should push a search query on the URL', () => {
      const spy = jest.fn();
      pushUrlQuery(spy, 'testing');
      expect(spy.mock.calls[0]).toMatchSnapshot();
    });
  });

});
