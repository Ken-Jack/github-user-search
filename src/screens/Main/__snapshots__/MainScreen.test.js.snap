// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Screen: MainScreen pushUrlQuery function should push a search query on the URL 1`] = `
Array [
  Object {
    "pathname": "/search",
    "search": "per_page=42&page=1&q=testing",
  },
]
`;

exports[`Screen: MainScreen when passed a component should render it in the default layout 1`] = `null`;
